cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -129,17 +161,38 @@ public void onRemoval(RemovalNotification<ArtifactDescriptor, Instantiators> not
     private final File pluginDir;
 
     private Instantiators(ArtifactDescriptor parentArtifactDescriptor) throws IOException {
-      this.parentClassLoader = artifactRepository.createArtifactClassLoader(parentArtifactDescriptor.getLocation());;
-      this.pluginInstantiatorMap = new HashMap<>();
+      this.parentClassLoader = artifactRepository.createArtifactClassLoader(parentArtifactDescriptor.getLocation());
+      this.pluginInstantiatorMap = new ConcurrentHashMap<>();
       this.pluginDir = DirUtils.createTempDir(stageDir);
     }
 
+    private boolean isArtifactSnapshotChanged(ArtifactDescriptor artifactDescriptor) {
+      if (artifactDescriptor.getArtifactId().getVersion().isSnapshot() &&
+        pluginInstantiatorMap.keySet().contains(artifactDescriptor)) {
+        for (ArtifactDescriptor descriptor : pluginInstantiatorMap.keySet()) {
```

Comment:
yeah, i thought it's okay as it's for snapshots, but missed the deletion case. will refactor.